src/lib/Templates.ml

```diff
@@ -0,0 +1,173 @@
+            match Hashtbl.find_opt param_map name with
+              | None -> Parsed_assignment (Parsed_scalar_update (name, id), rhs') :: tl'
+              | Some (Arg_name name') -> Parsed_assignment (Parsed_scalar_update (name', id), rhs') :: tl'
+              | Some _ -> failwith "[instantiate_instructions]: unexpected argument for template (expecting name)"
+        end
+        | Parsed_assignment (Parsed_indexed_update (name, index), rhs) ->
+            let rhs' = instantiate_boolean_expression param_map rhs in
+            Parsed_assignment (Parsed_indexed_update (name, index), rhs') :: tl'
+        | Parsed_instruction expr ->
+            let expr' = instantiate_boolean_expression param_map expr in
+            Parsed_instruction expr' :: tl'
+        | Parsed_for_loop ((name, id), left_expr, right_expr, up_down, body) -> begin
+            match Hashtbl.find_opt param_map name with
+              | None ->
+                  let left_expr' = instantiate_discrete_arithmetic_expression param_map left_expr in
+                  let right_expr' = instantiate_discrete_arithmetic_expression param_map right_expr in
+                  let body' = instantiate_instructions param_map body in
+                  Parsed_for_loop ((name, id), left_expr', right_expr', up_down, body') :: tl'
+              | Some v ->
+                  (* No need to isolate this one since the local variable is only scoped at `body` *)
+                  let left_expr' = instantiate_discrete_arithmetic_expression param_map left_expr in
+                  let right_expr' = instantiate_discrete_arithmetic_expression param_map right_expr in
+                  Hashtbl.remove param_map name;
+                  let body' = instantiate_instructions param_map body in
+                  Hashtbl.add param_map name v;
+                  Parsed_for_loop ((name, id), left_expr', right_expr', up_down, body') :: tl'
+        end
+        | Parsed_while_loop (cond, body) ->
+           let cond' = instantiate_boolean_expression param_map cond in
+           let body' = instantiate_instructions param_map body in
+           Parsed_while_loop (cond', body') :: tl'
+        | Parsed_if (cond, then_branch, else_branch_opt) ->
+            let cond' = instantiate_boolean_expression param_map cond in
+            let then_branch' = instantiate_instructions param_map then_branch in
+            let else_branch_opt' = Option.map (instantiate_instructions param_map) else_branch_opt in
+            Parsed_if (cond', then_branch', else_branch_opt') :: tl'
+        | Parsed_local_decl _ ->
+            raise (Exceptions.InternalError "[instantiate_instructions]: This point of code is unreachable") 
+
+let instantiate_transition param_map (guard, bloc, sync, loc_name) =
+  let guard' = instantiate_convex_predicate param_map guard in
+  let bloc' = instantiate_instructions param_map bloc in
+  (guard', bloc', sync, loc_name)
```

Comment:
No, as you said it is done in `rename_actions`. If an action was not in the template parameters, we have to make its name different in each instantiation of the template. This is the goal of the function `rename_actions`. It adds the name of the instantiation as a suffix in the name of the action (lines 125 and 137 of this file).

It would probably be shorter to integrate this modification in the `instantiate_locs` function... but I did this way to separate the concerns of the functions and to avoid passing the `user_name` to `instantiate_locs`